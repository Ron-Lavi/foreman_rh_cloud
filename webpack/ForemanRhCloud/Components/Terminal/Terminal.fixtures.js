export const logs = [
  'Generating...',
  'Hosts: 1000',
  'writing host 1/1000',
  'writing host 2/1000',
];

export const exitCode = 'Running pid 1111111';

export const props = { logs, exitCode };
