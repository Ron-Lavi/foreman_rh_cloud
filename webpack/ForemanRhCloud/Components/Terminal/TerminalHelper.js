export const isTerminalScrolledDown = (
  scrollHeight,
  scrollTop,
  terminalHeight,
  tolerance
) => scrollTop <= scrollHeight - terminalHeight - tolerance;
