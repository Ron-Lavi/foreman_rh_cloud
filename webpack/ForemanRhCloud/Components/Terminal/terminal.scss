.terminal {
  height: 200px;
  background-color: #222;
  padding: 10px 0;
  margin-bottom: 20px;
  overflow-y: scroll;
  overflow-x: hidden;

  p {
    font-family: monospace;
    font-size: 16px;
    color: #22da26;

    &.terminal_error {
      color: #f00;
    }
  }

  &::-webkit-scrollbar {
    width: 12px;
    height: 12px;
    background: #aaa;
  }

  &::-webkit-scrollbar-thumb {
    background: #222;
    border-radius: 6px;
    border: 3px solid transparent;
    background-clip: content-box;
  }
}
