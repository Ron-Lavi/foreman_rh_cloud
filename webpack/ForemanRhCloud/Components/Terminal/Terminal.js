import React from 'react';
import PropTypes from 'prop-types';
import { Grid, Spinner } from 'patternfly-react';
import { isEqual } from 'lodash';
import './terminal.scss';
import { isTerminalScrolledDown } from './TerminalHelper';

class Terminal extends React.Component {
  constructor(props) {
    super(props);
    this.terminal = React.createRef();
    this.state = {
      didUserScroll: false,
    };
  }

  componentDidMount() {
    this.scrollBottom();
  }

  componentDidUpdate({ logs: prevLogs }) {
    const { logs: currentLogs, autoScroll } = this.props;
    const { didUserScroll } = this.state;
    if (autoScroll && !didUserScroll && !isEqual(prevLogs, currentLogs)) {
      this.scrollBottom();
    }
  }

  handleScroll = e => {
    const { scrollTop, scrollHeight, offsetHeight } = e.target;
    const didUserScroll = isTerminalScrolledDown(
      scrollHeight,
      scrollTop,
      offsetHeight,
      100
    );
    this.setState({
      didUserScroll,
    });
  };

  scrollBottom = () => {
    const element = this.terminal.current;
    if (!element) {
      return;
    }
    const setHeightToBottom = () => {
      element.scrollTop = element.scrollHeight;
    };
    /** happens on tab switching when the terminal wasn't visible yet
     * and there was nothing to scroll, 250ms is enough to wait for the terminal to appear.
     */
    if (element.scrollHeight === 0) {
      setTimeout(setHeightToBottom, 250);
    } else {
      setHeightToBottom();
    }
  };

  render() {
    const { logs, error, exitCode } = this.props;
    let modifiedLogs = null;
    if (error !== null) {
      modifiedLogs = <p className="terminal_error">{error}</p>;
    } else if (Array.isArray(logs)) {
      modifiedLogs = logs.map((log, index) => <p key={index}>{log}</p>);
    } else {
      modifiedLogs = <p>{logs}</p>;
    }
    const exitCodeLowerCase = exitCode.toLowerCase();
    const loading =
      exitCodeLowerCase.indexOf('running') !== -1 ||
      exitCodeLowerCase.indexOf('restarting') !== -1;
    return (
      <Grid.Col sm={12}>
        <div
          className="terminal"
          ref={this.terminal}
          onScroll={this.handleScroll}
        >
          <Grid fluid>
            <Grid.Row>
              <Grid.Col sm={12}>
                {modifiedLogs}
                <Spinner loading={loading} inverse inline size="xs" />
              </Grid.Col>
            </Grid.Row>
          </Grid>
        </div>
      </Grid.Col>
    );
  }
}

Terminal.propTypes = {
  logs: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.string),
    PropTypes.string,
  ]),
  error: PropTypes.string,
  exitCode: PropTypes.string,
  autoScroll: PropTypes.bool,
};

Terminal.defaultProps = {
  logs: null,
  error: null,
  exitCode: '',
  autoScroll: true,
};

export default Terminal;
