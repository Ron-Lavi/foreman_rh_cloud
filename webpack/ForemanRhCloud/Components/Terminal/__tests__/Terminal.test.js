import React from 'react';
import { mount, testComponentSnapshotsWithFixtures } from '@theforeman/test';
import Terminal from '../Terminal';
import { props, logs } from '../Terminal.fixtures';

const fixtures = {
  'render without Props': {},
  'render with props': props,
};

describe('Terminal', () => {
  describe('rendering', () =>
    testComponentSnapshotsWithFixtures(Terminal, fixtures));

  it('handles terminal scroll on componentDidUpdate', () => {
    const wrapper = mount(<Terminal {...props} />);
    jest.spyOn(wrapper.instance(), 'scrollBottom');
    wrapper.setProps({ logs: [...logs, 'new-log'] });
    expect(wrapper.instance().scrollBottom).toBeCalled();
  });

  it('error should be displayed in terminal', () => {
    const modifiedProps = { ...props, error: 'some-error' };
    const wrapper = mount(<Terminal {...modifiedProps} />);
    expect(wrapper.find('p.terminal_error').exists()).toBeTruthy();
  });

  it('logs as a string instead of an array should be displayed', () => {
    const text = 'some-string-log';
    const modifiedProps = { ...props, logs: text };
    const wrapper = mount(<Terminal {...modifiedProps} />);
    expect(wrapper.find('.terminal p').text()).toEqual(text);
  });
});
