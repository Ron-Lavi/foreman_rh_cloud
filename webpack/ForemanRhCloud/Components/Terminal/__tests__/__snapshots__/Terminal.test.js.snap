// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Terminal rendering render with props 1`] = `
<Col
  bsClass="col"
  componentClass="div"
  sm={12}
>
  <div
    className="terminal"
    onScroll={[Function]}
  >
    <Grid
      bsClass="container"
      componentClass="div"
      fluid={true}
    >
      <Row
        bsClass="row"
        componentClass="div"
      >
        <Col
          bsClass="col"
          componentClass="div"
          sm={12}
        >
          <p
            key="0"
          >
            Generating...
          </p>
          <p
            key="1"
          >
            Hosts: 1000
          </p>
          <p
            key="2"
          >
            writing host 1/1000
          </p>
          <p
            key="3"
          >
            writing host 2/1000
          </p>
          <Spinner
            className=""
            inline={true}
            inverse={true}
            loading={true}
            size="xs"
          />
        </Col>
      </Row>
    </Grid>
  </div>
</Col>
`;

exports[`Terminal rendering render without Props 1`] = `
<Col
  bsClass="col"
  componentClass="div"
  sm={12}
>
  <div
    className="terminal"
    onScroll={[Function]}
  >
    <Grid
      bsClass="container"
      componentClass="div"
      fluid={true}
    >
      <Row
        bsClass="row"
        componentClass="div"
      >
        <Col
          bsClass="col"
          componentClass="div"
          sm={12}
        >
          <p />
          <Spinner
            className=""
            inline={true}
            inverse={true}
            loading={false}
            size="xs"
          />
        </Col>
      </Row>
    </Grid>
  </div>
</Col>
`;
